src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Extensions/IServiceCollectionExtensions.cs

```diff
@@ -120,16 +120,19 @@ public static void AddHoverServices(this IServiceCollection services)
         services.AddSingleton<IHoverInfoService, HoverInfoService>();
     }
 
-    public static void AddSemanticTokensServices(this IServiceCollection services)
+    public static void AddSemanticTokensServices(this IServiceCollection services, LanguageServerFeatureOptions featureOptions)
     {
-        services.AddHandlerWithCapabilities<SemanticTokensRangeEndpoint>();
+        if (!featureOptions.UseRazorCohostServer)
+        {
+            services.AddHandlerWithCapabilities<SemanticTokensRangeEndpoint>();
+            // Ensure that we don't add the default service if something else has added one.
+            services.TryAddSingleton<IRazorSemanticTokensInfoService, RazorSemanticTokensInfoService>();
+        }
```

Comment:
Given that there is no else case for this `if (!featureOptions.UseRazorCohostServer)` condition,

I wonder where is the cohost setup happening and why we don't add `IRazorSemanticTokensInfoService` for when `UseRazorCohostServer` is set to `true`.

Reply:
Cohosting uses MEF for DI, so the setup is all does via attributes and not by manually adding things to service collections